reference/FungibleToken.scilla

```diff
@@ -0,0 +1,394 @@
+    event e
+  end
+end
+
+(* @dev: Returns true if an address is an operator or default operator of a token_owner. *)
+(* @param operator:    Address of a potential operator.                                  *)
+(* @param token_owner: Address of a token_owner.                                         *)
+transition IsOperatorFor(token_owner: ByStr20, operator: ByStr20)
+  get_operator <- operators_map[token_owner][operator];
+  is_operator_approved = 
+    match get_operator with
+    | None => False
+    | Some status => status
+    end;
+  is_default_operator = is_default_operator f_eq operator default_operators;
+  is_revoked_operator <- exists revoked_default_operators[token_owner][operator];
+  is_default_operator_approved = let is_not_revoked_operator = 
+    negb is_revoked_operator in andb is_not_revoked_operator is_default_operator;
+  
+  is_approved = orb is_operator_approved is_default_operator_approved;
+  match is_approved with
+  | False =>
+    err = CodeNotAuthorized;
+    MakeErrorEvent err
+  | True =>
+    e = {_eventname : "IsOperatorForSuccess"; token_owner: token_owner; operator : operator};
+    event e
+  end
+end
+
+(* @dev: Sets amount as the allowance of spender over the caller’s tokens. Only token_owner allowed to invoke. *)
+(* param spender:      Address to be set as an approved_spender.                                               *)
+(* param amount:       Number of tokens to be set as allowance for the approved_spender.                       *)
+transition Approve(spender: ByStr20, amount: Uint128)
+  (* Checks if the _sender is approving himself *)
+  is_owner = builtin eq _sender spender;
+  match is_owner with
+  | True =>
+    err = CodeNotAuthorized;
+    MakeErrorEvent err
+  | False =>
+    ProcedureApprove spender amount
+  end
+end
+
+(* @dev: Returns the number of tokens an approved_spender is allowed to spend on behalf of the token_owner. *)
+(* param token_owner:  Address of a token_owner.                                                            *)
+(* param spender:      Address of the approved_spender.                                                     *)
+transition Allowance(token_owner: ByStr20, spender: ByStr20)
+  get_allowance <- allowances_map[token_owner][spender];
+  match get_allowance with
+  | None =>
+    (* approved_spender does not exists *)
+    err = CodeNotFound;
+    MakeErrorEvent err
+  | Some val =>
+    e = {_eventname : "Allowance"; token_owner: token_owner; spender : spender;  allowance : val};
+    event e
+  end
+end
+
+
+(* @dev: Returns the total amount of tokens in existence. *)
+transition TotalSupply()
+  current_supply <- total_tokens;
+  e = {_eventname : "TotalSupply"; total_supply : current_supply};
+  event e
+end
+
+(* @dev: Returns the amount of tokens owned by an address. *)
+transition BalanceOf(address: ByStr20)
+  get_bal <- balances_map[address];
+  balance = 
+    match get_bal with
+    | Some bal => bal
+    | None => Uint128 0
+    end;
```

Comment:
Candidate to be replace by a function call as in an analogous suggestion at the beginning of the contracts